File: sections/14_core_patterns_mixed_contest/problems/y_maximum_width_ramp/solution.py
import sys


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    n = data[0]
    values = data[1:]

    # Indices of strict prefix minima: all useful left endpoints.
    candidates: list[int] = []
    for i in range(n):
        if not candidates or values[i] < values[candidates[-1]]:
            candidates.append(i)

    answer = 0
    for j in range(n - 1, -1, -1):
        while candidates and values[candidates[-1]] <= values[j]:
            answer = max(answer, j - candidates.pop())

    print(answer)


if __name__ == "__main__":
    main()

File: sections/14_core_patterns_mixed_contest/problems/y_maximum_width_ramp/solution.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    int n;
    cin >> n;
    vector<long long> values(n);
    for (long long& value : values)
        cin >> value;

    // Only new prefix minima can be useful left endpoints.
    vector<int> candidates;
    for (int i = 0; i < n; ++i)
        if (candidates.empty() || values[i] < values[candidates.back()])
            candidates.push_back(i);

    int answer = 0;
    for (int j = n - 1; j >= 0; --j) {
        while (!candidates.empty() &&
               values[candidates.back()] <= values[j]) {
            answer = max(answer, j - candidates.back());
            candidates.pop_back();
        }
    }

    cout << answer << '\n';
}
